README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# tweetrmd

<!-- badges: start -->
<!-- badges: end -->

Easily embed Tweets anywhere R Markdown turns plain text into HTML.

## Installation

You can install the released version of **tweetrmd** from GitHub:

``` r
# install.packages("devtools")
devtools::install_github("gadenbuie/tweetrmd")
```

## Embed a Tweet

``` r
library(tweetrmd)
tweet_embed("https://twitter.com/alexpghayes/status/1211748406730706944")
```

<blockquote class="twitter-tweet" data-width="550" data-lang="en" data-dnt="true" data-theme="light"><p lang="en" dir="ltr">anybody have experience embedding tweets into <a href="https://twitter.com/hashtag/rmarkdown?src=hash&amp;ref_src=twsrc%5Etfw">#rmarkdown</a> documents *without using blogdown*?<a href="https://t.co/5kQUBh7j4g">https://t.co/5kQUBh7j4g</a></p>&mdash; alex hayes (@alexpghayes) <a href="https://twitter.com/alexpghayes/status/1211748406730706944?ref_src=twsrc%5Etfw">December 30, 2019</a></blockquote>

Or if you would rather use the screen name and status id.

``` r
tweet_embed(tweet_url("alexpghayes", "1211748406730706944"))
```

<blockquote class="twitter-tweet" data-width="550" data-lang="en" data-dnt="true" data-theme="light"><p lang="en" dir="ltr">anybody have experience embedding tweets into <a href="https://twitter.com/hashtag/rmarkdown?src=hash&amp;ref_src=twsrc%5Etfw">#rmarkdown</a> documents *without using blogdown*?<a href="https://t.co/5kQUBh7j4g">https://t.co/5kQUBh7j4g</a></p>&mdash; alex hayes (@alexpghayes) <a href="https://twitter.com/alexpghayes/status/1211748406730706944?ref_src=twsrc%5Etfw">December 30, 2019</a></blockquote>

In rich HTML outputs, the full embedded tweet is available and
interactive. Here, in GitHub-flavored markdown, only the content of the
tweet is seen.

## Embed many tweets

If you have several tweets you would like to embed at once, you can use
the following pattern to include add a vector of tweets to your
document. This works well when you want to include a [thread of
tweets](https://gist.github.com/gadenbuie/33c350458305f4423f30c1274be63b34).

``` r
thread <- c(
  "https://twitter.com/grrrck/status/1333804309272621060",
  "https://twitter.com/grrrck/status/1333804487148855300", 
  "https://twitter.com/grrrck/status/1333805092152123394"
)

htmltools::tagList(
  lapply(thread, tweet_embed, plain = TRUE)
)
```

<blockquote class="twitter-tweet" data-width="550" data-lang="en" data-dnt="true" data-theme="light"><p lang="en" dir="ltr">I&#39;ve got a new work laptop! I&#39;m going to try to track my setup process and the software and tools I install in this thread <a href="https://t.co/9X2qvHB3no">pic.twitter.com/9X2qvHB3no</a></p>&mdash; Garrick Aden-Buie (@grrrck) <a href="https://twitter.com/grrrck/status/1333804309272621060?ref_src=twsrc%5Etfw">December 1, 2020</a></blockquote>

<blockquote class="twitter-tweet" data-width="550" data-lang="en" data-dnt="true" data-theme="light"><p lang="en" dir="ltr">Step #1, wait... <a href="https://t.co/3533LZZQBt">pic.twitter.com/3533LZZQBt</a></p>&mdash; Garrick Aden-Buie (@grrrck) <a href="https://twitter.com/grrrck/status/1333804487148855300?ref_src=twsrc%5Etfw">December 1, 2020</a></blockquote>

<blockquote class="twitter-tweet" data-width="550" data-lang="en" data-dnt="true" data-theme="light"><p lang="en" dir="ltr">Oh wow, I really jumped the gun on this thread <a href="https://t.co/XpbzLTzStf">pic.twitter.com/XpbzLTzStf</a></p>&mdash; Garrick Aden-Buie (@grrrck) <a href="https://twitter.com/grrrck/status/1333805092152123394?ref_src=twsrc%5Etfw">December 1, 2020</a></blockquote>

(Note that I used `plain = TRUE` to embed each tweet [as
markdown](#embed-without-tracking).)

## Take a screenshot of a tweet

Screenshots are automatically embedded in R Markdown documents, or you
can save the screenshot as a `.png` or `.pdf` file. Uses the
[rstudio/webshot2](https://github.com/rstudio/webshot2) package.

``` r
tweet_screenshot(tweet_url("alexpghayes", "1211748406730706944"))
```

<img src="man/figures/README-screenshot-1.png" width="400px" />

## Just include a tweet in any R Markdown output format

When you want to include a tweet in multiple R Markdown formats, you can
use `include_tweet()`. It’s like `knitr::include_graphics()` but for
tweets. The function will automatically include the tweet as HTML in
HTML outputs, or as a screenshot in all others.

```` markdown
```{r tweet-from-dsquintana}
include_tweet("https://twitter.com/dsquintana/status/1275705042385940480")
```
````

<blockquote class="twitter-tweet" data-width="550" data-lang="en" data-dnt="true" data-theme="light"><p lang="en" dir="ltr">{bookdown} folks: I&#39;m trying to knit a PDF version of a HTML book that contains HTML elements (embedded tweets). <br><br>Is there a way to automatically take a screenshot of embedded tweets for PDF output? <br><br>Using the {webshot} package + PhantomJS didn&#39;t work...<a href="https://twitter.com/hashtag/Rstats?src=hash&amp;ref_src=twsrc%5Etfw">#Rstats</a></p>&mdash; Dan Quintana (@dsquintana) <a href="https://twitter.com/dsquintana/status/1275705042385940480?ref_src=twsrc%5Etfw">June 24, 2020</a></blockquote>

## Customize tweet appearance

Twitter’s [oembed
API](https://developer.twitter.com/en/docs/tweets/post-and-engage/api-reference/get-statuses-oembed)
provides a number of options, all of which are made available for
customization in `tweet_embed()` and `tweet_screenshot()`.

``` r
tweet_screenshot(
  tweet_url("alexpghayes", "1211748406730706944"),
  maxwidth = 300,
  hide_media = TRUE,
  theme = "dark"
)
```

<img src="man/figures/README-screenshot-customized-1.png" width="300px" />

## Embed without tracking

You can use `tweetrmd` to embed tweets in your documents and outputs
without including Twitter JavaScript or tracking. The easiest way is to
set `plain = TRUE` in `include_tweet()`. This will insert minimal HTML
for web outputs or convert the tweet text to markdown for non-web
outputs.

``` r
include_tweet(
  "https://twitter.com/dsquintana/status/1275705042385940480",
  plain = TRUE
)
```


    ```{=html}
    <blockquote class="twitter-tweet" data-width="550" data-lang="en" data-dnt="true" data-theme="light"><p lang="en" dir="ltr">{bookdown} folks: I&#39;m trying to knit a PDF version of a HTML book that contains HTML elements (embedded tweets). <br><br>Is there a way to automatically take a screenshot of embedded tweets for PDF output? <br><br>Using the {webshot} package + PhantomJS didn&#39;t work...<a href="https://twitter.com/hashtag/Rstats?src=hash&amp;ref_src=twsrc%5Etfw">#Rstats</a></p>&mdash; Dan Quintana (@dsquintana) <a href="https://twitter.com/dsquintana/status/1275705042385940480?ref_src=twsrc%5Etfw">June 24, 2020</a></blockquote>

    ```

Alternatively, you can choose to use `tweet_screenshot()` or
`tweet_markdown()` to embed all tweets in your documents.

## Caching tweets with memoization

Tweets are often deleted and re-running `tweet_embed()` or
`tweet_screenshot()` may fail or overwrite a previous screenshot of a
tweet. To avoid this, you can use the
[memoise](https://github.com/r-lib/memoise) package.

``` r
library(memoise)

tweet_cached <- memoise(tweet_embed, cache = cache_filesystem('.tweets'))
tweet_shot_cached <- memoise(tweet_screenshot, cache = cache_filesystem('.tweets'))
```

<sup>\*</sup>When memoising `tweet_screenshot()` you need to manually
save the file to a specific location. In the future my goal is for this
to be automatic.

------------------------------------------------------------------------

Note: When using `tweet_embed()`, you may need to add the following line
to your YAML header for strict markdown output formats.

``` yaml
always_allow_html: true
```


NEWS.md
# tweetrmd 0.0.10

# tweetrmd 0.0.9

* `tweet_embed()` and `tweet_url()` can now embed a tweet from just the status
  id of the tweet.


R/screenshot.R
#' Take or embed a screenshot of a tweet
#'
#' Takes a screenshot of the tweet that is automatically embedded in R Markdown
#' documents or that can be saved for use later.
#'
#' @examples
#' \dontrun{
#' tmpimg <- tempfile(fileext = "png")
#' tweet_screenshot(
#'   tweet_url = "https://twitter.com/alexpghayes/status/1211748406730706944",
#'   file = tmpimg
#' )
#' }
#'
#' @inheritParams tweet_embed
#' @param scale Scale the tweet for a better quality screenshot. Default is 2.
#' @param file A character string of output files. Can end with `.png` or `.pdf`.
#' @inheritDotParams tweet_embed
#' @family Tweet-embedding functions
#' @export
tweet_screenshot <- function(
  tweet_url,
  maxwidth = 550,
  scale = 2,
  file = NULL,
  link_color = NULL,
  ...
) {
  assert_string(tweet_url)
  requires_webshot2()

  html <- single_tweet_page(tweet_url, maxwidth, link_color, ...)

  tmpfile <- tempfile(fileext = ".html")
  htmltools::save_html(html, tmpfile, "transparent")

  webshot2::webshot(
    url = tmpfile,
    file = file %||% tempfile(fileext = ".png"),
    vwidth = maxwidth + 10,
    vheight = 2500,
    selector = "#screenshot-tweet",
    zoom = scale,
    expand = 0,
    delay = 0.5
  )
}

requires_webshot2 <- function(stop = TRUE) {
  has_webshot2 <- requireNamespace("webshot2", quietly = TRUE)
  if (!isTRUE(stop)) {
    has_webshot2
  } else if (!has_webshot2) {
    stop("`webshot2` is required: devtools::install_github('rstudio/webshot2')")
  } else {
    invisible(TRUE)
  }
}

single_tweet_page <- function(
  tweet_url,
  maxwidth = 550,
  link_color = NULL,
  ...
) {
  htmltools::tagList(
    htmltools::div(
      id = "screenshot-tweet",
      tweet_embed(tweet_url, maxwidth = maxwidth, ...)
    ),
    htmltools::tags$head(
      htmltools::tags$style(
        sprintf(
          read_pkg_file("tweet-screenshot-default.css"),
          link_color %||% "#3b94d9"
        )
      )
    )
  )
}


R/include_tweet.R
#' Include A Tweet in All R Markdown Formats
#'
#' Similar to [knitr::include_graphics()], but for tweets. In HTML documents,
#' the tweet is embedded using [tweet_embed()] and for all other documents types
#' a screen shot of the tweet is rendered and used [tweet_screenshot()]. If you
#' would rather that just the text of the tweet be included in non-HTML outputs,
#' use [tweet_embed()].
#'
#' @return An `htmltools::tagList()` to include a tweet in an HTML document, or
#'   a screen shot of the tweet for use in non-HTML outputs.
#'
#' @examples
#'
#' include_tweet("https://twitter.com/dsquintana/status/1275705042385940480")
#'
#' @inheritParams tweet_embed
#' @inheritDotParams tweet_embed
#' @family Tweet-embedding functions
#' @export
include_tweet <- function(tweet_url, plain = FALSE, ...) {
  if (!in_knitr() || knitr::is_html_output()) {
    return(tweet_embed(tweet_url, plain = plain, ...))
  }

  if (isTRUE(plain) || !requires_webshot2(stop = FALSE)) {
    knitr::asis_output(tweet_markdown(tweet_url, ...))
  } else {
    tweet_screenshot(tweet_url, ...)
  }
}

#' @describeIn include_tweet Return a tweet as plain markdown.
#' @export
tweet_markdown <- function(tweet_url, ...) {
  assert_string(tweet_url)
  bq <- tweet_blockquote(tweet_url, ...)
  html_to_markdown(bq)
}

html_to_markdown <- function(html, ...) {
  rmarkdown::pandoc_available(error = TRUE)
  tmpfile <- tempfile(fileext = ".html")
  tmpout <- tempfile(fileext = ".md")
  on.exit(unlink(c(tmpfile, tmpout)))
  writeLines(format(html), tmpfile)
  rmarkdown::pandoc_convert(tmpfile, from = "html", output = tmpout)
  x <- paste(readLines(tmpout), collapse = "\n")
  # strip twitter ?ref_src from urls
  gsub("(twitter[.]com.+?)([?][^)]+)", "\\1", x)
}


in_knitr <- function() {
  !is.null(knitr::current_input())
}


inst/styleTweet.js
(function styleTweets() {
  if (typeof twttr === 'undefined') {
    setTimeout(styleTweets, 0)
    return
  }
  if (!twttr.events) {
    setTimeout(styleTweets, 0)
    return
  }
  twttr.events.bind(
    'rendered',
    function (event) {
      // Fix dimensions of container div for webshot
      const st = document.getElementById('screenshot-tweet');
      if (event.target.shadowRoot) {
        // older versions of the widgets.js used shadow DOM
        event.target.shadowRoot.querySelector('.EmbeddedTweet').style.borderRadius = 0;
        const {width, height} = st.getBoundingClientRect();
        st.style.width = width + 'px';
        st.style.height = height + 'px';
      } else {
        // newer versions use an iframe and don't let us change the style
        setTimeout(function() {
          const {width, height} = st.querySelector('iframe').getBoundingClientRect();
          st.style.width = width + 'px';
          st.style.height = height + 'px';
        }, 200)
      }
    }
  );
})();


R/tweet.R
#' Embed a Tweet in R Markdown HTML outputs
#'
#' Uses Twitter's [oembed](https://publish.twitter.com) API to embed a
#' tweet in R Markdown HTML outputs.
#'
#' @references <https://developer.twitter.com/en/docs/tweets/post-and-engage/api-reference/get-statuses-oembed>
#' @family Tweet-embedding functions
#'
#' @examples
#' \dontrun{
#' # these are equivalent
#' tweet_embed("https://twitter.com/alexpghayes/status/1211748406730706944")
#'
#' tweet_embed(tweet_url("alexpghayes", "1211748406730706944"))
#'
#' tweet_embed("1211748406730706944")
#'
#' # Adding many tweets at once to a document takes a little bit more work
#' thread <- c(
#'   "https://twitter.com/grrrck/status/1333804309272621060",
#'   "https://twitter.com/grrrck/status/1333804487148855300",
#'   "https://twitter.com/grrrck/status/1333805092152123394"
#' )
#'
#' htmltools::tagList(
#'   lapply(thread, tweet_embed, plain = TRUE)
#' )
#' }
#'
#' @return The placeholder HTML for the tweet
#' @param tweet_url The URL for the tweet, or use [tweet_url()] to build the URL
#'   from a `screen_name` and `status_id`.
#' @param plain Embeds the tweet without including Twitter's JavaScript code.
#'   This means the tweets will be displayed as minimally styled HTML in
#'   [tweet_embed()] or [tweet_screenshot()]. In [include_tweet()],
#'   `plain = TRUE` will convert the tweet text into markdown for non-HTML
#'   outputs.
#'
#'   Note that for HTML outputs, you can't selectively choose whether or not a
#'   tweet is rendered in plain HTML if you include multiple tweet. Including
#'   one rich embedded tweet will trigger rich embedding of all tweets loaded
#'   on the page.
#' @param maxwidth The maximum width of a rendered Tweet in whole pixels. A
#'   supplied value under or over the allowed range will be returned as the
#'   minimum or maximum supported width respectively; the reset width value will
#'   be reflected in the returned `width` property. Note that Twitter does not
#'   support the oEmbed `maxheight` parameter. Tweets are fundamentally text,
#'   and are therefore of unpredictable height that cannot be scaled like an
#'   image or video. Relatedly, the oEmbed response will not provide a value for
#'   `height.` Implementations that need consistent heights for Tweets should
#'   refer to the `hide_thread` and `hide_media`` parameters below.
#' @param hide_media When set to `true`, `"t"`, or `1` links in a Tweet are not
#'   expanded to photo, video, or link previews.
#' @param hide_thread When set to `true`, `"t"`, or `1` a collapsed version of
#'   the previous Tweet in a conversation thread will not be displayed when the
#'   requested Tweet is in reply to another Tweet.
#' @param omit_script When set to `FALSE`, Twitter's remote script responsible
#'   for loading the Tweet widget will be used. Otherwise, the script included
#'   by this package will be used. If you are including multiple tweets on a
#'   page, it is recommended to use the default `FALSE`.
#' @param align Specifies whether the embedded Tweet should be floated left,
#'   right, or center in the page relative to the parent element.
#' @param related  A comma-separated list of Twitter usernames related to your
#'   content. This value will be forwarded to
#'   [Tweet action intents](https://developer.twitter.com/web/intents)
#'   if a viewer chooses to reply, like, or retweet the embedded Tweet.
#' @param lang Request returned HTML and a rendered Tweet in the specified
#'   [Twitter language supported by embedded Tweets](https://developer.twitter.com/web/overview/languages).
#' @param theme When set to `dark`, the Tweet is displayed with light text over
#'   a dark background.
#' @param link_color Adjust the color of Tweet text links with a
#'   [hexadecimal color value](https://en.wikipedia.org/wiki/Web_colors#Hex_triplet).
#' @param widget_type Set to `video` to return a Twitter Video embed for the
#'   given Tweet.
#' @param dnt When set to `true`, the Tweet and its embedded page on your site
#'   are not used for purposes that include
#'   [personalized suggestions](https://support.twitter.com/articles/20169421)
#'   and [personalized ads](https://support.twitter.com/articles/20170405).
#' @export
tweet_embed <- function(
  tweet_url,
  maxwidth = 550,
  plain = FALSE,
  hide_media = FALSE,
  hide_thread = FALSE,
  omit_script = TRUE,
  align = c("none", "left", "right", "center"),
  related = NULL,
  lang = "en",
  theme = c("light", "dark"),
  link_color = NULL,
  widget_type = NULL,
  dnt = TRUE
) {
  assert_string(tweet_url)
  if (!grepl("[^0-9]", tweet_url)) {
    tweet_url <- tweet_url(tweet_url)
  }

  if (plain) omit_script <- TRUE

  bq <- tweet_blockquote(
    tweet_url   = tweet_url,
    maxwidth    = maxwidth,
    hide_media  = hide_media,
    hide_thread = hide_thread,
    omit_script = omit_script,
    align       = align,
    related     = related,
    lang        = lang,
    theme       = theme,
    link_color  = link_color,
    widget_type = widget_type,
    dnt         = dnt
  )

  if (omit_script & !plain) {
    bq <- htmltools::tagList(html_dependency_twitter(), bq)
  }
  htmltools::browsable(bq)
}

#' @describeIn tweet_embed Create a URL for a tweet from a screen name and status id.
#' @param screen_name The user's screen name
#' @param status_id The tweet's status id
#' @export
tweet_url <- function(screen_name = NULL, status_id) {
  if ((!grepl("[^0-9]", as.character(screen_name)) && missing(status_id))) {
    status_id <- screen_name
    screen_name <- "status"
  } else if (!grepl("[^0-9]", as.character(status_id)) && is.null(screen_name)) {
    screen_name <- "status"
  }
  status_id <- as.character(status_id)
  stopifnot("status_id must contain only numbers" = grepl("[0-9]", status_id))
  sprintf("https://twitter.com/%s/status/%s", screen_name, status_id)
}

tweet_blockquote <- function(
  tweet_url,
  maxwidth = 550,
  hide_media = FALSE,
  hide_thread = FALSE,
  omit_script = TRUE,
  align = c("none", "left", "right", "center"),
  related = NULL,
  lang = "en",
  theme = c("light", "dark"),
  link_color = NULL,
  widget_type = NULL,
  dnt = TRUE,
  ...,
  result_only = FALSE
) {
  align <- match.arg(align)
  theme <- match.arg(theme)
  if (!is.null(widget_type) && !identical(widget_type, "video")) {
    stop("`widget_type` must be NULL or 'video'")
  }

  url <- httr::parse_url("https://publish.twitter.com/oembed")
  url$query <- list(
    url = tweet_url,
    maxwidth = maxwidth,
    hide_media = validate_true(hide_media),
    hide_thread = validate_true(hide_thread),
    omit_script = omit_script,
    align = align,
    related = related %||% paste(related, collapse = ","),
    lang = lang,
    theme = theme,
    link_color = link_color,
    widget_type = widget_type,
    dnt = validate_true(dnt),
    ...
  )

  url <- httr::build_url(url)
  res <- httr::GET(url)
  httr::stop_for_status(res)
  if (!grepl("application/json", res$headers$`content-type`)) {
    stop("Expected json response, got ", res$headers$`content-type`)
  }
  if (isTRUE(result_only)) return(res)
  res_txt <- httr::content(res, "text")
  res_json <- jsonlite::fromJSON(res_txt)
  htmltools::HTML(res_json$html)
}

html_dependency_twitter <- function() {
  htmltools::htmlDependency(
    name = "twitter-widget",
    version = "0.0.1",
    package = "tweetrmd",
    src = c(
      file = "tw",
      href = "https://platform.twitter.com/"
    ),
    script = "widgets.js",
    all_files = FALSE
  )
}

validate_true <- function(x) {
  if (isTRUE(x) || identical(tolower(x), "t") || x == 1) {
    TRUE
  } else FALSE
}

#' @describeIn tweet_embed A shortcut for the impatient.
#' @export
`%tweet%` <- function(screen_name, status_id) {
  tweet_url(screen_name, status_id)
}

